server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -663,6 +663,27 @@ public static String findV2Template(Metadata metadata, String indexName, @Nullab
             return null;
         }
 
+        // this is complex but if the index is not hidden in the create request but is hidden as the result of template application,
+        // then we need to exclude global templates
+        if (isHidden == null) {
```

Comment:
This can be simplified to something like this:
```java
if (isHidden == null) {
    matchedTemplates.entrySet().stream()
        .filter(entry -> IndexMetadata.INDEX_HIDDEN_SETTING.get(resolveSettings(metadata, entry.getValue())))
        // relies on the fact that default for INDEX_HIDDEN_SETTING is false
        .findAny()
        .ifPresent(indexTemplateV2ToNameEntry -> {
            matchedTemplates.keySet().removeIf(current -> current.indexPatterns().stream().anyMatch(Regex::isMatchAllPattern));
            if (matchedTemplates.containsKey(indexTemplateV2ToNameEntry.getKey()) == false) {
                throw new IllegalStateException("A global index V2 template [" + indexTemplateV2ToNameEntry.getValue() +
                    "] defined the index hidden setting, which is not allowed");
            }
        });
}
```

Reply:
On the second iteration on this I think there's a problem (fixed by proposal above): we can have 2 templates one with `INDEX_HIDDEN_SETTING=false` (explicitly set) and one with `INDEX_HIDDEN_SETTING=true`, `findFirst` can pick up `false` one and skip the whole block despite the fact that there's still one template with `true`